File: python/13-hitl/approval_flows.py
# Approval flows: clasificación de acciones por riesgo y gates de aprobación humana
#
# Cómo ejecutar:
#   make py SCRIPT=python/13-hitl/approval_flows.py
#
# Qué esperar:
#   El agente clasifica cada accion por nivel de riesgo (low/medium/high/critical).
#   Las acciones criticas esperan aprobacion humana via CLI.
#
# Variables de entorno:
#   MODEL — modelo a usar (default: claude-sonnet-4-6)

import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Callable, Optional
import os
import anthropic

cliente = anthropic.Anthropic()

# ─── Clasificación de riesgo ─────────────────────────────────────────────────

ACCIONES_ALTO_RIESGO = {
    "borrar_datos", "enviar_email_masivo", "transferencia_dinero",
    "modificar_cuenta_usuario", "desplegar_produccion", "revocar_accesos",
}

ACCIONES_MEDIO_RIESGO = {
    "escribir_produccion", "operacion_bulk", "cambiar_configuracion",
}

UMBRAL_REGISTROS = 100


def clasificar_accion(nombre: str, params: dict) -> str:
    if nombre in ACCIONES_ALTO_RIESGO:
        return "alto"
    if nombre in ACCIONES_MEDIO_RIESGO:
        return "medio"
    if params.get("registros_afectados", 0) > UMBRAL_REGISTROS:
        return "alto"
    if not params.get("reversible", True):
        return "alto"
    return "bajo"


def describir_impacto(nombre: str, params: dict) -> str:
    registros = params.get("registros_afectados", 0)
    tabla = params.get("tabla", "desconocida")
    if nombre == "borrar_datos":
        return f"Se borrarán {registros} registros de la tabla '{tabla}' en producción. Esta operación es irreversible."
    if nombre == "enviar_email_masivo":
        dest = params.get("destinatarios", 0)
        return f"Se enviará un email a {dest} usuarios. No puede deshacerse una vez enviado."
    return f"Acción '{nombre}' con parámetros: {json.dumps(params, ensure_ascii=False)}"


# ─── Cola de aprobaciones pendientes ─────────────────────────────────────────

@dataclass
class SolicitudAprobacion:
    id: str
    nombre_accion: str
    params: dict
    impacto: str
    timestamp: float
    expira_en: float  # unix timestamp
    decision: Optional[str] = None  # "aprobar" | "rechazar" | "modificar"
    params_modificados: Optional[dict] = None


_cola: dict[str, SolicitudAprobacion] = {}


def solicitar_aprobacion_sincrona(nombre: str, params: dict) -> dict:
    """Simula la interacción humana en CLI; en producción esto sería una UI/API."""
    impacto = describir_impacto(nombre, params)
    print(f"\n[APROBACIÓN REQUERIDA]")
    print(f"Acción: {nombre}")
    print(f"Impacto: {impacto}")
    print("Opciones: [a]probar / [r]echazar / [m]odificar")

    decision = input("Tu decisión: ").strip().lower()
    if decision.startswith("a"):
        return {"tipo": "aprobar", "params": params}
    if decision.startswith("m"):
        nuevos = input("Nuevos parámetros (JSON): ").strip()
        try:
            return {"tipo": "modificar", "params_modificados": json.loads(nuevos)}
        except json.JSONDecodeError:
            return {"tipo": "rechazar", "motivo": "parámetros modificados inválidos"}
    return {"tipo": "rechazar", "motivo": "rechazado por el usuario"}


def encolar_aprobacion(nombre: str, params: dict, ttl_horas: int = 4) -> str:
    """Encola para aprobación asíncrona (HOTL). Devuelve id de solicitud."""
    sol_id = str(uuid.uuid4())[:8]
    ahora = time.time()
    _cola[sol_id] = SolicitudAprobacion(
        id=sol_id,
        nombre_accion=nombre,
        params=params,
        impacto=describir_impacto(nombre, params),
        timestamp=ahora,
        expira_en=ahora + ttl_horas * 3600,
    )
    print(f"[COLA] Acción '{nombre}' encolada (id={sol_id}, expira en {ttl_horas}h)")
    return sol_id


# ─── Ejecutor con approval gate ───────────────────────────────────────────────

AprobadorFn = Callable[[str, dict], dict]


def ejecutar_herramienta_con_approval(
    nombre: str,
    params: dict,
    fn_herramienta: Callable,
    modo: str = "sincrono",  # "sincrono" (HITL) | "cola" (HOTL) | "auto"
) -> dict:
    nivel = clasificar_accion(nombre, params)

    if nivel == "bajo" or modo == "auto":
        resultado = fn_herramienta(nombre, params)
        print(f"[AUTO] {nombre}: {resultado}")
        return {"estado": "ejecutado", "resultado": resultado}

    if nivel == "medio" or modo == "cola":
        sol_id = encolar_aprobacion(nombre, params)
        return {"estado": "pendiente_revision", "id": sol_id}

    # nivel == "alto" → HITL bloqueante
    respuesta = solicitar_aprobacion_sincrona(nombre, params)

    if respuesta["tipo"] == "aprobar":
        resultado = fn_herramienta(nombre, respuesta["params"])
        return {"estado": "ejecutado", "resultado": resultado}
    if respuesta["tipo"] == "modificar":
        resultado = fn_herramienta(nombre, respuesta["params_modificados"])
        return {"estado": "ejecutado_modificado", "resultado": resultado}
    return {"estado": "rechazado", "motivo": respuesta.get("motivo", "")}


# ─── Agente con approval gate integrado ──────────────────────────────────────

HERRAMIENTAS = [
    {
        "name": "buscar_info",
        "description": "Busca información. Acción reversible y segura.",
        "input_schema": {
            "type": "object",
            "properties": {"query": {"type": "string"}},
            "required": ["query"],
        },
    },
    {
        "name": "borrar_datos",
        "description": "Borra registros de la base de datos. IRREVERSIBLE.",
        "input_schema": {
            "type": "object",
            "properties": {
                "tabla": {"type": "string"},
                "registros_afectados": {"type": "integer"},
            },
            "required": ["tabla", "registros_afectados"],
        },
    },
]


def _ejecutar_tool_real(nombre: str, params: dict) -> str:
    if nombre == "buscar_info":
        return f"Información encontrada para '{params['query']}': resultado simulado."
    if nombre == "borrar_datos":
        return f"[SIMULADO] Se habrían borrado {params['registros_afectados']} registros de '{params['tabla']}'."
    return f"Herramienta '{nombre}' no reconocida."


def agente_con_approval(tarea: str, modo_aprobacion: str = "sincrono") -> str:
    mensajes = [{"role": "user", "content": tarea}]

    for _ in range(10):
        respuesta = cliente.messages.create(
            model=os.environ.get("MODEL", "claude-sonnet-4-6"),
            max_tokens=1024,
            tools=HERRAMIENTAS,
            messages=mensajes,
        )

        mensajes.append({"role": "assistant", "content": respuesta.content})

        if respuesta.stop_reason == "end_turn":
            return next((b.text for b in respuesta.content if hasattr(b, "text")), "")

        if respuesta.stop_reason == "tool_use":
            tool_results = []
            for bloque in respuesta.content:
                if bloque.type != "tool_use":
                    continue
                resultado = ejecutar_herramienta_con_approval(
                    nombre=bloque.name,
                    params=bloque.input,
                    fn_herramienta=_ejecutar_tool_real,
                    modo=modo_aprobacion,
                )
                contenido = json.dumps(resultado, ensure_ascii=False)
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": bloque.id,
                    "content": contenido,
                })
            mensajes.append({"role": "user", "content": tool_results})

    return "[max iteraciones]"


if __name__ == "__main__":
    print("=== Clasificación de riesgo ===")
    tests = [
        ("buscar_info", {"query": "usuarios activos"}),
        ("escribir_produccion", {"tabla": "users", "registros_afectados": 50}),
        ("borrar_datos", {"tabla": "users", "registros_afectados": 847}),
    ]
    for nombre, params in tests:
        nivel = clasificar_accion(nombre, params)
        print(f"  {nombre}: {nivel}")

    print("\n=== Agente con approval (modo auto — sin interacción) ===")
    resultado = agente_con_approval(
        "Busca información sobre usuarios activos en el último mes.",
        modo_aprobacion="auto",
    )
    print(f"Resultado: {resultado[:200]}")

File: typescript/13-hitl/approval_flows.ts
// Cómo ejecutar: make ts SCRIPT=typescript/13-hitl/approval_flows.ts
import Anthropic from "@anthropic-ai/sdk";
import * as readline from "readline";
import { randomUUID } from "crypto";

const cliente = new Anthropic();

const ACCIONES_ALTO_RIESGO = new Set([
  "borrar_datos", "enviar_email_masivo", "transferencia_dinero",
  "modificar_cuenta_usuario", "desplegar_produccion", "revocar_accesos",
]);

const ACCIONES_MEDIO_RIESGO = new Set([
  "escribir_produccion", "operacion_bulk", "cambiar_configuracion",
]);

const UMBRAL_REGISTROS = 100;

function clasificarAccion(nombre: string, params: Record<string, unknown>): string {
  if (ACCIONES_ALTO_RIESGO.has(nombre)) return "alto";
  if (ACCIONES_MEDIO_RIESGO.has(nombre)) return "medio";
  if ((params["registros_afectados"] as number ?? 0) > UMBRAL_REGISTROS) return "alto";
  if (params["reversible"] === false) return "alto";
  return "bajo";
}

function describirImpacto(nombre: string, params: Record<string, unknown>): string {
  const registros = params["registros_afectados"] as number ?? 0;
  const tabla     = params["tabla"] as string ?? "desconocida";
  if (nombre === "borrar_datos") {
    return `Se borrarán ${registros} registros de la tabla '${tabla}' en producción. Esta operación es irreversible.`;
  }
  if (nombre === "enviar_email_masivo") {
    const dest = params["destinatarios"] as number ?? 0;
    return `Se enviará un email a ${dest} usuarios. No puede deshacerse una vez enviado.`;
  }
  return `Acción '${nombre}' con parámetros: ${JSON.stringify(params)}`;
}

interface SolicitudAprobacion {
  id:                 string;
  nombre_accion:      string;
  params:             Record<string, unknown>;
  impacto:            string;
  timestamp:          number;
  expira_en:          number;
  decision:           string | null;
  params_modificados: Record<string, unknown> | null;
}

const _cola: Record<string, SolicitudAprobacion> = {};

async function solicitarAprobacionSincrona(
  nombre: string,
  params: Record<string, unknown>
): Promise<{ tipo: string; params?: Record<string, unknown>; params_modificados?: Record<string, unknown>; motivo?: string }> {
  const impacto = describirImpacto(nombre, params);
  console.log(`\n[APROBACIÓN REQUERIDA]`);
  console.log(`Acción: ${nombre}`);
  console.log(`Impacto: ${impacto}`);
  console.log("Opciones: [a]probar / [r]echazar / [m]odificar");

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  return new Promise((resolve) => {
    rl.question("Tu decisión: ", (decision) => {
      decision = decision.trim().toLowerCase();

      if (decision.startsWith("a")) {
        rl.close();
        resolve({ tipo: "aprobar", params });
        return;
      }

      if (decision.startsWith("m")) {
        rl.question("Nuevos parámetros (JSON): ", (nuevos) => {
          rl.close();
          try {
            resolve({ tipo: "modificar", params_modificados: JSON.parse(nuevos.trim()) });
          } catch {
            resolve({ tipo: "rechazar", motivo: "parámetros modificados inválidos" });
          }
        });
        return;
      }

      rl.close();
      resolve({ tipo: "rechazar", motivo: "rechazado por el usuario" });
    });
  });
}

function encolarAprobacion(nombre: string, params: Record<string, unknown>, ttlHoras = 4): string {
  const solId = randomUUID().slice(0, 8);
  const ahora = Date.now() / 1000;
  _cola[solId] = {
    id:                 solId,
    nombre_accion:      nombre,
    params,
    impacto:            describirImpacto(nombre, params),
    timestamp:          ahora,
    expira_en:          ahora + ttlHoras * 3600,
    decision:           null,
    params_modificados: null,
  };
  console.log(`[COLA] Acción '${nombre}' encolada (id=${solId}, expira en ${ttlHoras}h)`);
  return solId;
}

type HerramientaFn = (nombre: string, params: Record<string, unknown>) => string;

async function ejecutarHerramientaConApproval(
  nombre: string,
  params: Record<string, unknown>,
  fnHerramienta: HerramientaFn,
  modo = "sincrono"
): Promise<Record<string, unknown>> {
  const nivel = clasificarAccion(nombre, params);

  if (nivel === "bajo" || modo === "auto") {
    const resultado = fnHerramienta(nombre, params);
    console.log(`[AUTO] ${nombre}: ${resultado}`);
    return { estado: "ejecutado", resultado };
  }

  if (nivel === "medio" || modo === "cola") {
    const solId = encolarAprobacion(nombre, params);
    return { estado: "pendiente_revision", id: solId };
  }

  const respuesta = await solicitarAprobacionSincrona(nombre, params);

  if (respuesta.tipo === "aprobar") {
    const resultado = fnHerramienta(nombre, respuesta.params!);
    return { estado: "ejecutado", resultado };
  }
  if (respuesta.tipo === "modificar") {
    const resultado = fnHerramienta(nombre, respuesta.params_modificados!);
    return { estado: "ejecutado_modificado", resultado };
  }
  return { estado: "rechazado", motivo: respuesta.motivo ?? "" };
}

const HERRAMIENTAS: Anthropic.Tool[] = [
  {
    name:        "buscar_info",
    description: "Busca información. Acción reversible y segura.",
    input_schema: {
      type:       "object",
      properties: { query: { type: "string" } },
      required:   ["query"],
    },
  },
  {
    name:        "borrar_datos",
    description: "Borra registros de la base de datos. IRREVERSIBLE.",
    input_schema: {
      type:       "object",
      properties: {
        tabla:               { type: "string" },
        registros_afectados: { type: "number" },
      },
      required: ["tabla", "registros_afectados"],
    },
  },
];

function ejecutarToolReal(nombre: string, params: Record<string, unknown>): string {
  if (nombre === "buscar_info")  return `Información encontrada para '${params["query"]}': resultado simulado.`;
  if (nombre === "borrar_datos") return `[SIMULADO] Se habrían borrado ${params["registros_afectados"]} registros de '${params["tabla"]}'.`;
  return `Herramienta '${nombre}' no reconocida.`;
}

async function agenteConApproval(tarea: string, modoAprobacion = "sincrono"): Promise<string> {
  const mensajes: Anthropic.MessageParam[] = [{ role: "user", content: tarea }];

  for (let i = 0; i < 10; i++) {
    const respuesta = await cliente.messages.create({
      model:      process.env["MODEL"] ?? "claude-sonnet-4-6",
      max_tokens: 1024,
      tools:      HERRAMIENTAS,
      messages:   mensajes,
    });

    mensajes.push({ role: "assistant", content: respuesta.content });

    if (respuesta.stop_reason === "end_turn") {
      return respuesta.content.find((b): b is Anthropic.TextBlock => b.type === "text")?.text ?? "";
    }

    if (respuesta.stop_reason === "tool_use") {
      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const bloque of respuesta.content) {
        if (bloque.type !== "tool_use") continue;
        const resultado = await ejecutarHerramientaConApproval(
          bloque.name,
          bloque.input as Record<string, unknown>,
          ejecutarToolReal,
          modoAprobacion
        );
        toolResults.push({
          type:        "tool_result",
          tool_use_id: bloque.id,
          content:     JSON.stringify(resultado),
        });
      }
      mensajes.push({ role: "user", content: toolResults });
    }
  }

  return "[max iteraciones]";
}

async function main() {
  console.log("=== Clasificación de riesgo ===");
  const tests: [string, Record<string, unknown>][] = [
    ["buscar_info",        { query: "usuarios activos" }],
    ["escribir_produccion", { tabla: "users", registros_afectados: 50 }],
    ["borrar_datos",       { tabla: "users", registros_afectados: 847 }],
  ];
  for (const [nombre, params] of tests) {
    const nivel = clasificarAccion(nombre, params);
    console.log(`  ${nombre}: ${nivel}`);
  }

  console.log("\n=== Agente con approval (modo auto — sin interacción) ===");
  const resultado = await agenteConApproval(
    "Busca información sobre usuarios activos en el último mes.",
    "auto"
  );
  console.log(`Resultado: ${resultado.slice(0, 200)}`);
}

main();
